initial_commit to pysteps-nwp-importers

RubenImhoff opened the pull request:
This PR is a first step to a separate nwp importers plugin, which can be used in the pysteps repo. It resolves issue #218 in the pysteps repo and was discussed in the `blending-v1` branch in the pysteps repo.

To do:

- [x] Make sure tests run automatically
- [x] Check if tests run and adjust (I had some problems with the BoM NWP importer via the plugin)
- [x] Download files from `pysteps-data-master`, needed for tests

RubenImhoff committed:
initial_commit

RubenImhoff committed:
minor docstring changes

RubenImhoff committed:
add automatic tests and black

RubenImhoff committed:
automatic tests v2

RubenImhoff commented:
Alright, from here I might need some help. As far as I understand CI, I have tried to make some commits, but not sure if this is what it should be. 

Two other points of attention:

- BoM importer and test fail, but I haven't figured out why. Perhaps it is related to `@postprocess_import()`. 
- Right now all test data is downloaded, but in principle we only need the nwp folder.

aperezhortal committed:
Remove pysteps imports

- Remove the imports from the pysteps packages to avoid circular dependencies.
- Create a MissingOptionalDependency class.
- Remove the postprocessing import decorator since this is already applied when the plugin is imported.

aperezhortal commented:
> BoM importer and test fail, but I haven't figured out why. Perhaps it is related to @postprocess_import().

You are right, there was a problem with the cookie-cutter template :sweat_smile:  (I will fix it today). The postprocess_import decorator is not needed since it is being applied during the plugin discovery. 

Commit 85e5085 removes the decorator, and also other imports from pysteps to avoid any potential issue with circular dependencies.

I will try to take a look at the CI workflow today.

RubenImhoff commented:
Wonderful! I only still keep getting an error when trying to import the BoM data. Let's see if you have the same, it might be on my side, it gives a `Windows fatal exception: access violation`.

RubenImhoff committed:
Minor changes to make code independent of pystepsrc

aperezhortal committed:
Fix and refactor tests

- Remove pysteps imports
- Organized the importer tests into a single class
- Add fixtures for each importer
- Add script to download the test data from github

aperezhortal committed:
Remove template comments

aperezhortal committed:
Update CI tests

aperezhortal committed:
Remove old references to pysteps_data

aperezhortal committed:
Use absolute data paths to test data

aperezhortal committed:
Add importers discovery test

aperezhortal committed:
Add pysteps as dependency for CI/tests

aperezhortal committed:
Update tox rules

aperezhortal committed:
Fix test issues iwth installed packages

- Move the tests folder inside the package
- Move the download data module inside the test directory
- Force the tests to download the data if it does not exist
- Update the CI/Test workflow

aperezhortal committed:
Fix tox instructions

aperezhortal committed:
Remove unnecessary dependencies/comments

aperezhortal commented:
The tests are up and running :rocket: 

But....there is a problem with windows and py3.9 that seems related to mamba. For some reason, it is trying to install pysteps=1.0.0...

RubenImhoff commented:
Great work, @aperezhortal! Yes, looks like it has trouble installing mamba - no idea why.. 
Good to see the other tests pass! Locally they do too, except for the BoM importer, which oddly enough still returns a `RuntimeError: NetCDF: HDF error` on my Python 3.9 installation.

RubenImhoff committed:
Add time stamps to metadata

RubenImhoff committed:
add time stamp to tests

dnerini commented:
Great work @RubenImhoff and @aperezhortal! 

Going forward we should think whether it would make more sense to separate the importers into individual plugins, instead of having them all toghether. Moreover, ideally the importer plugins will be maintained directly by the data owners (ie. BOM, KNMI, RMI) and hosted on their Github page, but let's make one step at the time ;-)

More importantly: currently the only way to install the plugin is with the source code, but I don't see why we shouldn't publish it on pypi? This will also facilitate our CI workflows with the pysteps base package, since it will only take one `pip install` to get the importer. It should be rather easy to implement the relevant github workflow and update the documentation. If you agree, I'll make the needed changes.

dnerini commented on docs/installation.rst:
```diff
@@ -0,0 +1,40 @@
+The latest development version of pysteps_nwp_importers can be installed using
+pip by running in a terminal::
+
+    pip install git+https://github.com/YOUR_GITHUB_USERNAME/pysteps_nwp_importers
```
```suggestion
    pip install git+https://github.com/pySTEPS/pysteps-nwp-importers
```

dnerini commented on docs/installation.rst:
```diff
@@ -0,0 +1,40 @@
+Stable release
+--------------
+
+To install pysteps_nwp_importers, run this command in your terminal:
```
Isn't the name of the package "pysteps-nwp-importers"?

RubenImhoff replied:
Correct! I started of with pysteps_nwp_importers and then changed it to be more consistent with the other repositories. I'll change it.

dnerini commented on docs/installation.rst:
```diff
@@ -0,0 +1,40 @@
+Installation
+============
+
+.. The following installation instructions are recommended if the plugin is uploaded to pypi.
```
Let's keep this only if the plugins if published on pypi

dnerini commented on docs/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+# Additional requeriments related to the documentation build only
```
```suggestion
# Additional requirements related to the documentation build only
```

dnerini commented on requirements.txt:
```diff
@@ -0,0 +1,13 @@
+numpy
+xarray
+netCDF4
+pysteps
```
```suggestion
pysteps>=1.4,<2.0
```

I suggest to explicitly state the minimum (1.4 introduced the plugins) and maximum (2.0 will likely change the interface to the importers) versions here.

dnerini commented on requirements_dev.txt:
```diff
@@ -0,0 +1,16 @@
+numpy
+xarray
+netCDF4
+pysteps
```
```suggestion
pysteps>=1.4,<2.0
```

(see above)

dnerini commented on setup.py:
```diff
@@ -0,0 +1,42 @@
+with open("README.rst") as readme_file:
+    readme = readme_file.read()
+
+requirements = ["numpy", "xarray", "dask", "netCDF4"]
```
```suggestion
requirements = ["numpy", "xarray", "dask", "netCDF4", "pysteps>=1.4,<2.0"]
```

dnerini commented on setup.py:
```diff
@@ -0,0 +1,42 @@
+        "License :: OSI Approved :: BSD License",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
```
```suggestion
        "Programming Language :: Python :: 3.8",
        "Programming Language :: Python :: 3.9",        
```

dnerini commented on setup.py:
```diff
@@ -0,0 +1,42 @@
+            "import_rmi_nwp=pysteps_nwp_importers.importer_rmi_nwp:import_rmi_nwp",
+        ]
+    },
+    version="1.6",
```
I don't think this should refer to the version from the base package, but rather use its own. Should we go for 0.1.0?

RubenImhoff replied:
Good point, I agree!

dnerini commented on ci/ci_test_env.yml:
```diff
@@ -0,0 +1,15 @@
+  - mamba
+  - numpy
+  - netCDF4
+  - pysteps
```
```suggestion
  - pysteps>=1.4,<2.0
```

RubenImhoff committed:
Update docs/installation.rst

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff committed:
Update docs/requirements.txt

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff committed:
Update requirements.txt

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff committed:
Update requirements_dev.txt

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff committed:
Update setup.py

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff committed:
Update setup.py

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff committed:
Update ci/ci_test_env.yml

Co-authored-by: Daniele Nerini <daniele.nerini@meteoswiss.ch>

RubenImhoff commented:
> Going forward we should think whether it would make more sense to separate the importers into individual plugins, instead of having them all toghether. Moreover, ideally the importer plugins will be maintained directly by the data owners (ie. BOM, KNMI, RMI) and hosted on their Github page, but let's make one step at the time ;-)

Hmm yeah, it would make sense to do it separately, but I can imagine the pySTEPS base repo filling up quickly with different importers if we don't directly move it to local Github pages. Should we just continue like this for now and keep this as a next step? (no strong opinion here, whatever works!)

> More importantly: currently the only way to install the plugin is with the source code, but I don't see why we shouldn't publish it on pypi? This will also facilitate our CI workflows with the pysteps base package, since it will only take one `pip install` to get the importer. It should be rather easy to implement the relevant github workflow and update the documentation. If you agree, I'll make the needed changes.

Good idea!

RubenImhoff committed:
update docs

RubenImhoff committed:
update installation docs

RubenImhoff committed:
Change version, author name and email in setup.py

dnerini commented:
Mmmmh the python 3.9 windows test keeps failing. A bit weird since the same setup seems to work fine in https://github.com/pySTEPS/pysteps/pull/257 ...

Not sure what we could try next, maybe using conda to setup mamba as in https://github.com/pySTEPS/pysteps/pull/256 ?

RubenImhoff commented:
It is a bit weird, indeed.. 

> Not sure what we could try next, maybe using conda to setup mamba as in pySTEPS/pysteps#256 ?

Yeah, why not, let's try that!

dnerini committed:
Keep only one README file

dnerini committed:
Use markdown for long_description

dnerini committed:
Install pysteps as installation requirement with pip

dnerini committed:
Add publish workflow

dnerini committed:
Preinstall pysteps on MacOS

dnerini committed:
Rename master to main

dnerini committed:
Format with black

dnerini commented:
The tests on windows pass when pysteps is installed with pip, which could be an indication that something's wrong with the relative conda-forge package (did we remove windows-py39 from conda-forge? @aperezhortal ?).

Unforuntaly using pip now causes the (usual) issue with gcc on macOS, which I fixed with a not-so-nice workaround (preinstall step, hardcode version min-max). 

Not super elegant, but good enough for me at this point, you agree?

dnerini committed:
Use README.rst and remove comments from the template

dnerini committed:
Minor changes to doc

RubenImhoff commented:
Great work, @dnerini! Ready to merge?

aperezhortal commented:
@dnerini, we are not building the conda package for python3.9 in windows: https://github.com/conda-forge/pysteps-feedstock/blob/master/.azure-pipelines/azure-pipelines-win.yml 

But I don't remember why. Probably we remove it because some dependencies were missing back then.

We can merge this PR and later on update the workflow when the py39 package is built for windows.

aperezhortal commented:
> Going forward we should think whether it would make more sense to separate the importers into individual plugins, instead of having them all together. Moreover, ideally, the importer plugins will be maintained directly by the data owners (ie. BOM, KNMI, RMI) and hosted on their Github page, but let's make one step at the time ;-)

I agree with the idea of transferring the maintenance responsibilities to the data owners since custom importers are kind of outside of the scope of pysteps. 

We can move forward with this PR to have a functional example. But before would not upload it to pypi at this moment if the importer will be maintained by another organization or individual. They can upload it to pypi.

dnerini commented:
> We can move forward with this PR to have a functional example. But before would not upload it to pypi at this moment if the importer will be maintained by another organization or individual. They can upload it to pypi.

Mh yes, i see your point. Ok let's not publish to pypi for now, meaning that we'll have to install from source for the example gallery (which is totally ok)

dnerini commented:
update: with d8848f0f53b97b3a9d3bb39138d56950a5d53af2 we removed the option to install from pypi.
